new SleepFinishedTimeEvent

Cheaterpaul opened the pull request:
The new event is fired when all players are fully asleep and the time should be set to day.
Now you are able to set a custom time for wake up.

Cheaterpaul committed:
new SleepFinishedTimeEvent

Alpvax commented on patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch:
```diff
@@ -73,15 +73,17 @@
+-            this.func_72877_b(l - l % 24000L);
 +            long l = this.func_72820_D() + 24000L;
-             this.func_72877_b(l - l % 24000L);
++            this.func_72877_b(net.minecraftforge.event.ForgeEventFactory.onSleepFinished(this, l - l % 24000L));
```
Would this cause a problem if the resulting time is before the current time? (I'm not 100% sure how time is handled in MC). If that is the case, would it not make more sense to just return a time-of-day, rather than an absolute time?

Alpvax replied:
It was more of a question than a request for change.
Can you still sleep for less than a night with the new approach? If not it would be best to revert that commit.

Might be better to wait for someone who knows more about it to comment before making further changes.

Cheaterpaul replied:
now it adds the time to the current daytime (adds min 0 ticks, max 1 day)

^^should be better and safer than my first approach

Cheaterpaul committed:
now adding time

instead of setting the dayTime, now the time get added to the daytime if its between 0-24000

maxanier commented on src/main/java/net/minecraftforge/event/world/SleepFinishedTimeEvent.java:
```diff
@@ -0,0 +1,59 @@
+ *
+ * setWakeUpTime(wakeUpTime) sets a new time that will be added to the dayTime.<br>
+ */
+@Cancelable
```
Any reason this is cancelable? The result is used even if it is canceled. 
Only difference is that if a mod cancels it, the remaining mods won't get the event. Which makes it first wins instead of last wins

Orange1861 commented:
Once the canceled issue gets fixed, then I think that this could be merged.

Cheaterpaul committed:
remove cancelable

LexManos commented on patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch:
```diff
@@ -73,15 +73,17 @@
-             this.func_72877_b(l - l % 24000L);
+-            this.func_72877_b(l - l % 24000L);
++            long i = net.minecraftforge.event.ForgeEventFactory.onSleepFinished(this);
++            this.func_72877_b(i < 0L || i > 24000L ? func_72820_D() + 24000L - this.func_72820_D() % 24000L : func_72820_D() + i);
```
The logic here is complex. Why not just pass in a 'target time' and not let anyone set the result time to anything less then the current time?

So setTime(onSleepFinished(l - l %24000L))

Cheaterpaul committed:
moved checks to Event

LexManos commented on src/main/java/net/minecraftforge/event/world/SleepFinishedTimeEvent.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    public boolean setTimeAddition(long newTimeIn)
+    {
+        if (getWorld().getWorld().getDayTime() > newTimeIn)
```
getWorld().getWorld()? Also, why not just pass in minTime to compare here?

Cheaterpaul committed:
passing minTime

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6043) <br/>All committers have signed the CLA.

maxanier reviewed (approved):
Looks good
